lib/rust/ensogl/core/src/system/gpu/shader/compiler.rs

```diff
@@ -0,0 +1,466 @@
+        on_ready: F,
+    ) -> JobHandler {
+        self.rc.borrow_mut().submit(input, on_ready)
+    }
+
+    /// Run the compiler. This should be run on every frame.
+    pub fn run(&self, time: animation::TimeInfo) {
+        self.rc.borrow_mut().run(time)
+    }
+}
+
+impl CompilerData {
+    fn new(context: &native::ContextWithExtensions) -> Self {
+        let dirty = false;
+        let context = context.clone();
+        let jobs = default();
+        let performance = web::window.performance_or_panic();
+        let logger = Logger::new("Shader Compiler");
+        Self { dirty, context, jobs, performance, logger }
+    }
+
+    fn submit<F: 'static + FnOnce(shader::Program)>(
+        &mut self,
+        input: shader::Code,
+        on_ready: F,
+    ) -> JobHandler {
+        self.dirty = true;
+        let strong_handler = JobHandler::new();
+        let handler = strong_handler.downgrade();
+        let on_ready = Box::new(on_ready);
+        let job = Job { input, handler, on_ready };
+        self.jobs.compile.push(job);
+        strong_handler
+    }
+
+    #[profile(Debug)]
+    fn run(&mut self, time: animation::TimeInfo) {
+        if self.dirty {
+            self.run_khr_completion_check_jobs();
+            while self.dirty {
+                match self.run_step() {
+                    Ok(made_progress) => {
+                        if !made_progress {
+                            break;
+                        }
+                        let now = (self.performance.now() as f32).ms();
+                        let current_frame_time =
+                            now - time.animation_loop_start - time.since_animation_loop_started;
+                        if current_frame_time > FRAME_TIME_THRESHOLD {
+                            let msg1 =
+                                "Shaders compilation takes more than the available frame time.";
+                            let msg2 = "To be continued in the next frame.";
+                            debug!(self.logger, "{msg1} {msg2}");
+                            break;
+                        }
+                    }
+                    Err(err) => {
+                        if self.context.is_context_lost() {
+                            break;
+                        }
+                        let err_msg = err.blocking_report(&self.context);
+                        error!(self.logger, "{err_msg}");
+                    }
+                }
+            }
+        }
+    }
+
+    /// Runs the next compiler job if there is any left and if it will not cause too many jobs being
+    /// run in parallel. The result [`bool`] indicates if the call to this function did any
+    /// progress.
+    #[profile(Detail)]
+    fn run_step(&mut self) -> Result<bool, Error> {
+        let did_progress = |_| Ok(true);
+        let no_progress = Ok(false);
+        let jobs = &self.jobs;
+        if self.current_parallel_job_count() < MAX_PARALLEL_COMPILE_JOBS {
```

Comment:
We could gate only `run_next_compile_job`, since none of the others increase the number of outstanding jobs